### PR TITLE
core: Add initial 3rd party integration

### PR DESCRIPTION
**Pull request by patrickelectric**
This follows our documentation described in the wiki page
To test:
create the following file in a path: `register_service`
content:
```js
{
    "name":"Extension name",
    "description":"A small description about the service",
    "icon":"mdi-alpha-p-box",
    "company":"Blue Robotics",
    "version":"1.2.0",
    "webpage":"https://github.com/bluerobotics/BlueOS-docker",
    "api":""
 }
```
start simple python http server:
`python3 -m http.server 9992`

docker: patrickelectric/blueos-core:add_3rd_party_integration

### CODE REVIEW

**Comment by rafaellehmkuhl** [comment]
The service discovery appears to be working correctly, with the extensions menu opening when I serve a new extension, but the extension name apparently is not being caught:

image

**Comment by patrickelectric** [comment]
> The service discovery appears to be working correctly, with the extensions menu opening when I serve a new extension, but the extension name apparently is not being caught:
> 
> image

Did you restart the core docker or helper service ? It caches the last result.

**Comment by rafaellehmkuhl** [comment]
> > The service discovery appears to be working correctly, with the extensions menu opening when I serve a new extension, but the extension name apparently is not being caught:
> > image
> 
> Did you restart the core docker or helper service ? It caches the last result.

My steps:

1. Created the `register_service` file, equal to your example (with name equal to "extension name")
2. Served it with python `http.server`
3. Everything working as intended!
4. Stop the server
5. Extensions menu disappear, great!
6. Edited the `register_service` file with another extension name
7. Served it again
8. Menu reappears, but menu name is as before, without the changes

Looking at the `web_services` request, it appears there's some caching going on, as it is getting the same info as before the changes, for all the fields:

image

**Comment by rafaellehmkuhl** [comment]
> > The service discovery appears to be working correctly, with the extensions menu opening when I serve a new extension, but the extension name apparently is not being caught:
> > image
> 
> Did you restart the core docker or helper service ? It caches the last result.

Humn, I see now.

Can we prevent this caching somehow?
It seems counter-intuitive that we need to refresh the helper service/core container to catch the extension changes.

(this can be addressed in another PR, for time concerns)

**Comment by patrickelectric** [comment]
> > > The service discovery appears to be working correctly, with the extensions menu opening when I serve a new extension, but the extension name apparently is not being caught:
> > > image
> > 
> > 
> > Did you restart the core docker or helper service ? It caches the last result.
> 
> Humn, I see now.
> 
> Can we prevent this caching somehow? It seems counter-intuitive that we need to refresh the helper service/core container to catch the extension changes.
> 
> (this can be addressed in another PR, for time concerns)

This is done by our time cache decorator, it should update after some time.

**Comment by rafaellehmkuhl** [comment]
> 

It doesn't appear to be it. The helper cache is 1 minute. It's been more than 10 minute here, and it's still the same.

Will restart core to see how it goes.

**Comment by rafaellehmkuhl** [comment]
> > 
> 
> It doesn't appear to be it. The helper cache is 1 minute. It's been more than 10 minute here, and it's still the same.
> 
> Will restart core to see how it goes.

When core is restarted, the changes are catch.

File: core/services/helper/main.py [hunk 1]
```diff
@@ -52,6 +63,15 @@ def detect_service(port: int) -> ServiceInfo:
         if not info.valid:
             return info
 
+        # Check for service description metadata
+        try:
+            with requests.get(f"http://127.0.0.1:{port}/register_service", timeout=0.2) as response:
+                if response.status_code == http.HTTPStatus.OK:
+                    info.metadata = ServiceMetadata.parse_obj(response.json())
+        except Exception:
```
Comment (Williangalvani) [inline core/services/helper/main.py:71]: shouldn't pylint catch this as a broad-exception?

we should at least log the exception here

**Review by Williangalvani — APPROVED** [review]
working fine here, but I think we should log that exception.

image

File: core/services/helper/main.py (hunk 1 above)
Comment (patrickelectric) [reply core/services/helper/main.py:71]: > shouldn't pylint catch this as a broad-exception?
> 
> we should at least log the exception here

We have it disabled: https://github.com/patrickelectric/companion-docker/blob/8ab788c945a1f656c2deedaaddf0096fb29fab21/pyproject.toml#L55

But I agree that we should log the issue.

Comment (rafaellehmkuhl) [reply core/services/helper/main.py:71]: It should. Strange. And I can't find any `# pylint: ignore` blocks anywhere.

**Comment by patrickelectric** [comment]
> working fine here, but I think we should log that exception.
> 
> image

nice icon for dvl :rofl:

File: core/frontend/src/types/helper.ts [hunk 2]
```diff
@@ -1,7 +1,18 @@
...
+}
+
 export interface Service {
     valid: boolean
     title: string
     documentation_url: string
     versions: Array<string>
     port: number
+    metadata?: ServiceMetadata
 }
```
Comment (rafaellehmkuhl) [inline core/frontend/src/types/helper.ts:18]: Regarding this, how should one creating a service behave in respect to `name` vs `title`, `valid`, `webpage` vs `api` vs `documentation_url`, for example?

Comment (patrickelectric) [reply core/frontend/src/types/helper.ts:18]: Service is a structure that we use to help us easily identify services with the helper scraping logic. the metadata is the real information provided by the running service. 
We should rely on the metadata to provide all the necessary information for the frontend. Service was only a quick way to make it work without having this concept finished.
Remember that this is still not a supported thing and we can rework everything around this concept before have official support. The PR only allows the metadata access for people and services that are willing to have it working right now and accept the API break in the future.

**Review by rafaellehmkuhl — APPROVED** [review]

**Pull request merged**